### DISCUSSION TITLE
How to get notified when the application exits?

### DISCUSSION THREAD

**Discussion by myfix16** [microsoft/WindowsAppSDK#3516] [Q&A] [answered]
I wonder, in Windows App SDK, is it possible to get notified when the app exits? I know we can listen to the `MainWindow.Close` event. But what if users kill the app via task manager, or the app is terminated by the OS for some reason (say, shutdown)? Is there any way to get notified in these situations?

Appreciate any idea.

**Comment by castorix** [comment, marked as answer]
On shutdown, you can intercept _WM_QUERYENDSESSION_ 
Task Manager sends _WM_SYSCOMMAND_ with _SC_CLOSE_ to the main window, but if you intercept it, it kills the process...
(also calls _RtlReportSilentProcessExit_)

**Reply by myfix16** [reply]
@castorix  Thanks for your reply. I tried intercepting `WM_SYSCOMMAND` with `SC_CLOSE`. However, when I killed the app using the task manager, the message is not captured.  

Below is partial code in `MainWindow.xaml.cs`:
```
    internal ILogger<MainWindow>? Logger;

    public delegate int SUBCLASSPROC(IntPtr hWnd, uint uMsg, IntPtr wParam, IntPtr lParam, IntPtr uIdSubclass, uint dwRefData);

    [DllImport("Comctl32.dll", SetLastError = true)]
    public static extern bool SetWindowSubclass(IntPtr hWnd, SUBCLASSPROC pfnSubclass, uint uIdSubclass, uint dwRefData);

    [DllImport("Comctl32.dll", SetLastError = true)]
    public static extern int DefSubclassProc(IntPtr hWnd, uint uMsg, IntPtr wParam, IntPtr lParam);

    public const int WM_CLOSE = 0x0010;
    public const int WM_QUERYENDSESSION = 0x11;
    public const int WM_ENDSESSION = 0x16;
    public const int WM_SYSCOMMAND = 0x0112;
    public const int SC_CLOSE = 0xF060;


    IntPtr hWnd = IntPtr.Zero;
    private SUBCLASSPROC SubClassDelegate;

    public MainWindow()
    {
        InitializeComponent();

        Logger = App.GetService<ILogger<MainWindow>>();

        Content = null;
        Title = "AppDisplayName".Localized();

        hWnd = WinRT.Interop.WindowNative.GetWindowHandle(this);
        SubClassDelegate = new SUBCLASSPROC(WindowSubClass);
        bool bReturn = SetWindowSubclass(hWnd, SubClassDelegate, 0, 0);
    }

    private int WindowSubClass(IntPtr hWnd, uint uMsg, IntPtr wParam, IntPtr lParam, IntPtr uIdSubclass, uint dwRefData)
    {
        switch (uMsg)
        {
            case WM_CLOSE:
                Logger?.LogInformation("App receives WM_CLOSE");
                break;
            case WM_ENDSESSION:
                Logger?.LogInformation("App receives WM_ENDSESSION");
                break;
            case WM_SYSCOMMAND when wParam == SC_CLOSE:
                Logger?.LogInformation("App receives WM_SYSCOMMAND - SC_CLOSE");
                break;
        }
        return DefSubclassProc(hWnd, uMsg, wParam, lParam);
    }
```
Is there any problem with my code?

**Reply by DarranRowe** [reply]
If you are going through the details page in task manager, the End Task button uses [TerminateProcess](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-terminateprocess). I think it eventually uses TerminateProcess on the processes page if the process doesn't respond.
As the name may imply, this is unconditional, completely one sided and the process being terminated has no chance to execute any more code.

**Reply by castorix** [reply]
I get a Beep when I kill the process with Task Manager, on Process page (not in Details as _DarranRowe_ said),  (on Windows 10 22H2, maybe the "_TaskManagerWindow_" class, that I test to check if it is on foreground, is different on Windows 11, must be checked with Spy++...)
Then the process is killed of course...
On XP, _NtTerminateProcess_ could be hooked, but I did not test on Windows 10 (API Hooks are complex)...

        case WM_SYSCOMMAND:
        {
            if (((int)wParam & (int)0xFFF0) == SC_CLOSE)
            {
                if (GetForegroundWindow() == FindWindow("TaskManagerWindow", null))
                {
                    Console.Beep(5000, 10);
                    return 0;
                } 
            }
        }
        break;

with :

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern IntPtr FindWindow(string lpClassName, string lpWindowName);

        [DllImport("User32.dll", SetLastError = true, CharSet = CharSet.Unicode)]
        public static extern IntPtr GetForegroundWindow();

**Reply by myfix16** [reply]
@castorix Can you provide the full file where 
```
    case WM_SYSCOMMAND:
    {
        if (((int)wParam & (int)0xFFF0) == SC_CLOSE)
        {
            if (GetForegroundWindow() == FindWindow("TaskManagerWindow", null))
            {
                Console.Beep(5000, 10);
                return 0;
            } 
        }
    }
    break;
```
resides? So that I can try it on my computer. Thanks.

**Reply by castorix** [reply]
> @castorix Can you provide the full file where
> ...
> resides? So that I can try it on my computer. Thanks.

This is just the code inside WindowSubClass, like in your sample
